src/onnx/onnx_parser.cpp

```diff
@@ -224,28 +224,42 @@ int64_t onnx_parser::get_opset_version(const onnx::ModelProto& model)
 
 void onnx_parser::parse_graph(module* mod, const onnx::GraphProto& graph)
 {
+    std::unordered_map<std::string, instruction_ref> mod_insts;
     for(auto&& f : graph.initializer())
     {
-        instructions[f.name()] = mod->add_literal(parse_tensor(f));
+        // backup instructions in parent mod
+        mod_insts[f.name()] = mod->add_literal(parse_tensor(f));
     }
 
     for(auto&& input : graph.input())
     {
         const std::string& name = input.name();
         // input not in initializer_data, so it is a real input
-        if(!contains(instructions, name))
+        if(!contains(mod_insts, name))
         {
             std::vector<std::size_t> dims;
             if(map_input_dims.count(name) > 0)
             {
                 dims = map_input_dims.at(name);
             }
 
-            shape s            = parse_type(input.type(), dims);
-            instructions[name] = mod->add_parameter(name, s);
+            shape s         = parse_type(input.type(), dims);
+            mod_insts[name] = mod->add_parameter(name, s);
         }
     }
 
+    // backup parameters/literals with the same names as that in
+    // the parent module
+    std::unordered_map<std::string, instruction_ref> pmod_insts_bkup;
+    for(auto& ins_p : mod_insts)
+    {
+        if(contains(instructions, ins_p.first))
+        {
+            pmod_insts_bkup[ins_p.first] = instructions[ins_p.first];
+        }
+        instructions[ins_p.first] = ins_p.second;
+    }
```

Comment:
What is this doing? Is this to handle parameter shadowing? I thought shadowing is not permitted by ONNX spec from [Nodes](https://github.com/onnx/onnx/blob/master/docs/IR.md#nodes):

> The graph MUST use single static assignment for all node outputs, which means that all node output names MUST be unique within a graph. In the case of a nested subgraph, a node output name MUST be distinct from the names from the outer scopes that are visible in the nested subgraph.

Is that an incorrect interpretation? Do you see onnx models that are shadowing params and constants?

Reply:
We have another scenario that subgraph and parent graph have parameters with the same name. In this case, when we parse the subgraph, the instruction (stored in the mem var ```instructions```  corresponding to the parameter in the subgraph will overwrite that of the parent graph. After parsing subgraph returns, if we do not restore the parameter of the parent graph and and the parent graph uses this parameter as input for some other operators, inputs of the operator are the parameter of the subgraph.

In this scenario, the name is not output of any node, so the above specification does not apply, and there is no specification saying parent graph and subgraph cannot use the same name as parameter names. The only thing I read from the notes is:

>In nested subgraphs used as attribute values, users MUST NOT use the same name as both a subgraph initializer and subgraph input unless the corresponding op's specification explicitly allows it.

but this does not apply, either. 

I got the example with the above scenario some time ago online, but I cannot find it anymore. Also, when I created the onnx file from the script, no error is reported.

Reply:
@pfultz2 Changed accordingly and added a unit test for the run loop. Could you please take a look again? Thanks